src/fsharp/FSharp.Core/prim-types.fs

```diff
@@ -1505,12 +1636,179 @@ namespace Microsoft.FSharp.Core
                     override iec.Equals(x:obj,y:obj) = GenericEqualityObj true iec (x,y)  // ER Semantics
                     override iec.GetHashCode(x:obj) = raise (InvalidOperationException (SR.GetString(SR.notUsedForHashing))) }
 
+            let canUseDefaultEqualityComparer (ty:Type) =
+                let processed = System.Collections.Generic.HashSet ()
+
+                let rec recurse idx (types:array<Type>) =
+                    if idx = types.Length then true
+                    else
+                        let ty = get types idx
+                        if not (processed.Add ty) then
+                            recurse (idx+1) types
+                        else
+                            let isValidGenericType ifNotType fullname =
+                                if not (ty.IsGenericType && ty.GetGenericTypeDefinition().FullName.Equals fullname)
+                                then ifNotType
+                                else recurse 0 (ty.GetGenericArguments ())
+                    
+                            // avoid any types that need special handling in GenericEqualityObj
+                            true 
+                            && ty.IsSealed  // covers enum and value types
+                            && not ty.IsArray
+                            && not (ty.Equals typeof<String>)
+                            && not (ty.Equals typeof<Decimal>)
+                            && not (ty.Equals typeof<float>)
+                            && not (ty.Equals typeof<float32>)
+                            && isValidGenericType true "System.Nullable`1"
+                            && not (typeof<IStructuralEquatable>.IsAssignableFrom ty
+                                    && not (false
+                                            || isValidGenericType false "System.ValueTuple`1"
```

Comment:
Why do you need to special case `ValueTuple`?  There is no corresponding `ValueTuple` case in `GenericEqualityObj`?  

Have you put this in because `ValueTuple` is known to respect structural checking, i.e. if `T1` and `T2` support default comparison then `ValueTuple<T1,T2>`?  Would it make sense to add `Result` and `ValueOption` to this table too?

Reply:
I was confused about whether this code applies to enum and value types.  I see `// covers enum and value types` but it seems it applies to reference types as well?    Could you add more comments in this please?

Also the double negative `not (..... isValidGenericType false "System.ValueTuple`1")` is really confusing and something feels wrong here, How can we have both 

     not (isValidGenericType false "System.ValueTuple`1" || .... )

and

    isValidGenericType true "System.Nullable`1"

The recursive call to `isValidGenericType` seems to be used in both a positive and negative sense in these different cases.  Simplify/clarify the code please so it's really easy to think about, since this routine is really important to have correct.

Reply:
`ValueTuples` have nice implementation of `IEquatable<>.Equals` which utilizes `EqualityComparer<'Tn>.Default` so if all the `'Tns` are safe (which we check recursively) then we get a really quick comparison. Unfortuantely standard tuple in its `IEquatable<>.Equals` method uses EqualityComparer<obj>.Default on all Tns, so really doesn't help. I haven't looked at Result and ValueOption.

The second parameter in isValidGenericType  is if the _root type_ doesn't match (i.e. is the type Nullable? no, then return true so check the next `&&` condition - or is the type a ValueTuple? no, then return false to check the next `||` condition...), otherwise the result of the function is true if all it's generic arguments are good or false if one of them fails. 

The `&& not (false` followed by the `||`s can be read as _unless ..._ 

I did play around a bit with this whole logic. I'm not sure if there is a way to have it so that on first glance it reads well, but I think the combination of the wonky formatting :-) and maybe a couple of comments might help? I don't know, if you think you can make it much easier then feel free to mess around with it. I just think the first way that it clicks in your head is the easiest - which might be different for different people so not sure of the gain of changing it? But happy to do so.

But I will fix up some comments.